File: 61-rotate-list/61-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null){
            return head;
        }
        ListNode cur = head;
        int listLen = 1;
        while (cur.next != null){
            cur = cur.next;
            listLen ++;
        }
        
        int rotation = k % listLen;
        if (rotation == 0){
            return head;
        }
        
        // we made a circle
        cur.next = head;
        int roation = k;
        if (k>listLen){
            rotation = k % listLen;
        }
        
        int nodeIdx = listLen - rotation;
        
        cur = head;
        ListNode newHead = head;
        while (nodeIdx != 1){
            nodeIdx --;
            cur = cur.next;
        }
        
        newHead = cur.next;
        cur.next = null;
        return newHead;
        
        
        
    }
}

File: 61-rotate-list/61-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        
        list_length = 0
        cur = head
        if head is None or head.next is None:
            return head
        
        while cur.next is not None:
            list_length += 1
            cur = cur.next
        
        list_length += 1
        
        if k == list_length or k == 0:
            return head
        
        cur.next = head
        break_idx = list_length - (k % list_length)
        
        cur = head
        for i in range(1, break_idx):
            cur = cur.next
        
        head = cur.next
        cur.next = None
        
        return head
            
            
        
        